File: 0207-course-schedule/0207-course-schedule.py
class Node:
    def __init__(self):
        self.indegree = 0
        self.outNodes = []

class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        from collections import defaultdict, deque
        graph = defaultdict(Node)
        
        countEdges = 0
        for prereq in prerequisites:
            course, dependentCourse = prereq[0], prereq[1]
            graph[dependentCourse].outNodes.append(course)
            graph[course].indegree += 1
            countEdges +=1
            
        notDependentCourses = deque()
        for index, item in graph.items():
            if item.indegree == 0:
                notDependentCourses.append(index)
        
        removedEdges = 0
        while notDependentCourses:
            notDepCourse = notDependentCourses.pop()
            for outNode in graph[notDepCourse].outNodes:
                graph[outNode].indegree -= 1
                removedEdges += 1
                if graph[outNode].indegree == 0:
                    notDependentCourses.append(outNode)
        
        return (removedEdges == countEdges)
                
        

File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> inDegreeV(numCourses);
        vector<vector<int>> adjMat(numCourses, vector<int>());
        for(auto & n : prerequisites)
        {
            adjMat[n[1]].push_back(n[0]);
            inDegreeV[n[0]]++;
        }
        queue<int> q;
        for(int i = 0; i < numCourses; ++i) {
            
            if(inDegreeV[i] == 0) q.push(i); // Push the course number (index), not the in-degree.
        }

        vector<int> topoOrderedList;
        while(!q.empty())
        {
            auto topE = q.front();
            q.pop();
            topoOrderedList.push_back(topE);
            for(auto & l : adjMat[topE])
            {
                --inDegreeV[l];
                if(inDegreeV[l] == 0) q.push(l);
            }
        }
        return topoOrderedList.size() == numCourses;
    }
};